lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -1171,9 +1196,91 @@ struct RvalueExprVisitor : public ExprVisitor {
         expr.subroutine);
   }
 
+  /// Get both the actual `this` argument of a method call and the required
+  /// class type.
+  std::pair<mlir::Value, moore::ClassHandleType>
+  getMethodReceiverTypeHandle(const slang::ast::CallExpression &expr) {
+
+    moore::ClassHandleType handleTy;
+    mlir::Value thisRef;
+
+    // Qualified call: t.m(...), extract from thisClass.
+    if (const slang::ast::Expression *recvExpr = expr.thisClass()) {
+      thisRef = context.convertRvalueExpression(*recvExpr);
+      if (!thisRef)
+        return {};
+
+      handleTy = dyn_cast<moore::ClassHandleType>(thisRef.getType());
+      if (!handleTy) {
+        mlir::emitError(loc)
+            << "receiver of method '" << expr.getSubroutineName()
+            << "' must be class<...>, got " << thisRef.getType();
+        return {};
+      }
+    } else {
+      // Unqualified call inside a method body: try using implicit %this.
+      thisRef = context.getImplicitThisRef();
+      if (!thisRef) {
+        mlir::emitError(loc) << "method '" << expr.getSubroutineName()
+                             << "' called without an object";
+        return {};
+      }
+      handleTy = dyn_cast<moore::ClassHandleType>(thisRef.getType());
+      if (!handleTy) {
+        mlir::emitError(loc)
+            << "implicit 'this' must be class<...>, got " << thisRef.getType();
+        return {};
+      }
```

Comment:
Can you add error tests cases that trigger each of these errors? I suspect that most of these cannot be triggered because Slang prevents such an AST from being formed, in which case I'd turn these into asserts.